src/lib.rs

```diff
@@ -64,31 +93,61 @@ fn parse_effective_cpus(raw_cpus: String) -> Result<c_int> {
 }
 
 fn effective_cpus_count(base_cg: PathBuf) -> Result<c_int> {
-    let mut cur = Some(base_cg.as_path());
+    let raw_cpus = read_cg_controller(base_cg, "cpuset.cpus.effective")?;
+    parse_effective_cpus(raw_cpus)
+}
 
-    while let Some(cg) = cur {
-        let path: PathBuf = [
-            "/sys/fs/cgroup",
-            &cg.to_string_lossy(),
-            "cpuset.cpus.effective",
-        ]
-        .iter()
-        .collect();
-
-        if let Ok(v) = fs::read_to_string(&path) {
-            return parse_effective_cpus(v);
-        }
+fn parse_cfs_quota_as_cpus(cpus_max: String) -> Result<c_int> {
+    let parts: Vec<&str> = cpus_max.split_whitespace().collect();
+    if parts.len() != 2 {
+        return Err(Errno { errno: EINVAL });
+    }
+    let quota: c_int = parts[0].parse().map_err(|_| Errno { errno: EINVAL })?;
+    let period: c_int = parts[1].parse().map_err(|_| Errno { errno: EINVAL })?;
 
-        cur = cg.parent()
+    if period == 0 {
+        return Err(Errno { errno: EINVAL }); // Avoid division by zero
     }
 
-    Err(Errno { errno: ENOENT })
+    Ok((quota / period) as c_int)
+}
+
+fn cpu_count_from_quota(base_cg: PathBuf) -> Result<c_int> {
+    let cpus_max = read_cg_controller(base_cg, "cpu.max")?;
+    parse_cfs_quota_as_cpus(cpus_max)
+}
+
+fn get_affinity_cpu_count() -> Result<c_int> {
+    let pid = Pid::from_raw(0);
+    let cpuset = sched_getaffinity(pid)?;
```

Comment:
Can you describe what sched_getaffinity() does that isn't covered by cpusets? I suppose it takes into account isolcpus?

I'm asking because I think we're more likely to be successful in landing this elsewhere if we have less external dependencies.

Reply:
They are just totally different mechanisms. cpusets and cpu affinity masks are distinct, its just that the scheduler takes them both into account.

You can see this by setting an affinity mask and seeing that cpuset.cpus.effective doesnt notice or by setting a cpuset and seeing that sched_getaffinity doesnt notice.

But we should take both into account since they can both constrain the scheduler.

Reply:
I see, makes sense, thanks. I wonder if we can call this syscall directly, vs. having an external dependency. but we already have a library size problem, so it's probably fine for now.

Reply:
Looks to me like cpusets are a strict superset of affinity: https://github.com/torvalds/linux/blob/610347effc2ecb5ededf5037e82240b151f883ab/kernel/sched/core.c#L8348-L8349

Does that mean we can get rid of the cgroup querying code (for cpusets at least) all together?

Reply:
Is that because taskset looks at cpusets (via sched_getaffinity) before it tries to set? IIUC, you could manually sched_setaffinity() and that would work, assuming that set was a subset of your cpuset. (What I'm really asking is: can we delete some of my code in favor of yours here?)